micrometer-meter-provider/src/main/java/io/opentelemetry/contrib/metrics/micrometer/internal/instruments/AbstractUpDownCounter.java

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.contrib.metrics.micrometer.internal.instruments;
+
+import io.micrometer.core.instrument.FunctionCounter;
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.contrib.metrics.micrometer.internal.state.InstrumentState;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+abstract class AbstractUpDownCounter extends AbstractInstrument {
+  private final Map<Attributes, AtomicDoubleCounter> counterMap = new ConcurrentHashMap<>();
+
+  protected AbstractUpDownCounter(InstrumentState instrumentState) {
+    super(instrumentState);
+  }
+
+  protected final void add(Attributes attributes, double value) {
+    counterMap.computeIfAbsent(attributesOrEmpty(attributes), this::createCounter).increment(value);
+  }
+
+  protected final void record(double value, Attributes attributes) {
+    counterMap.computeIfAbsent(attributesOrEmpty(attributes), this::createCounter).set(value);
+  }
+
+  private AtomicDoubleCounter createCounter(Attributes attributes) {
+    AtomicDoubleCounter counter = new AtomicDoubleCounter();
+    FunctionCounter.builder(name(), counter, AtomicDoubleCounter::current)
```

Comment:
`FunctionCounter` is supposed to be a monotonically increasing counter; for the `UpDownCounter` it's probably better to use a Micrometer `Gauge`.

Reply:
You're right, it looked like Micrometer didn't enforce that FunctionCounter was monotonic but the underlying MeterRegistry implementation might enforce it or make assumptions.  I've switched back to Gauge for correctness.